### PR TITLE
Dropped support for Python 2

### LINKED ISSUE #1929: Proposal: Drop Python 2 support in v0.20.0
At this point, most frameworks in our ecosystem have dropped Python 2 support, so it makes sense for us to consider doing the same.  

Already, a number of our more recent features (elastic mode, synchronized batch normalization) are Python 3 only.

We still officially maintain support for Python 2 for all legacy code paths, but there are a number of reasons to consider dropping this:

1. Running Python 2 unit tests cost money to run in the cloud, and since we have nearly 1:1 Python 2 to Python 3 test coverage, this nearly doubles our cloud bill for CPU tests.
2. As new versions of Python continue to be released, it becomes a challenge to just maintain mutual compatibility between Python 3.6 (our earliest supported version of Python 3) and newer versions (python 3.8, which recently changes the default multiprocessing run mode, see #1891). 
3. Maintaining feature parity becomes a challenge, as a number of basic things supported natively in Python 3 require elaborate workarounds for Python 2 (for example, multiprocessing barrier).

Additionally, users who need to use Python 2 can continue to use older versions of Horovod.  As TensorFlow and PyTorch should not be releasing new Python 2 versions of their libraries, this shouldn't be an issue for users who are stuck using Python 2.

What does dropping Python 2 support mean?

1. No more Python 2 tests in Buildkite.
2. No more Python 2 Docker images or PyPi packages.
3. No more system version checks to perform different logic for Python 2 vs Python 3.
4. No more dependency on `six`, replace all usages with native Python 3 equivalents.
5. No more doc references to Python 2, or unit test exclusions based on Python 2.

TSC members, please review and provide your feedback:

@alsrgv @nvcastet @romerojosh @EnricoMi @karakusc @apeforest

### PR DESCRIPTION
**Pull request by tgaddair**
Fixes #1929.

- Removed Python 2 tests from Buildkite.
- Removed references to Python 2 in Docker images.
- Removed system version checks that perform different logic for Python 2 vs Python 3.
- Removed dependency on `six` and replaced all usages with native Python 3 equivalents.
- Removed doc references to Python 2 and added Python 3 to install requirements.
- Removed unit test exclusions for Python 2.
- Removed Python 2 from supported versions in setup.

### CODE REVIEW

**Comment by EnricoMi** [comment]
Do we need to mention Python3 in docs somewhere?

File: .buildkite/gen-pipeline.sh [hunk 1]
```diff
@@ -298,7 +285,7 @@ run_spark_integration() {
   # Horovod Spark Estimator tests
   if [[ ${test} != *"tf1_1_0"* && ${test} != *"tf1_6_0"* && ${test} != *"torch0_"* && ${test} != *"mpich"* && ${test} != *"oneccl"* ]]; then
     if [[ ${test} != *"tf2"* && ${test} != *"tfhead"* ]]; then
-      if [[ ! (  ${test} == *"py2"* && ${test} == *"gloo"* && ${test} != *"openmpi-gloo"* ) ]]; then
+      if [[ ! (  ${test} == *"gloo"* && ${test} != *"openmpi-gloo"* ) ]]; then
```
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:288]: now that there are only two conditions left, the overall negation should be applied to simplify the expression:

    ${test} != *"gloo"* || ${test} == *"openmpi-gloo"*

File: .buildkite/gen-pipeline.sh [hunk 2]
```diff
@@ -319,7 +306,7 @@ run_spark_integration() {
       fi
     fi
 
-    if [[ ! (  ${test} == *"py2"* && ${test} == *"gloo"* && ${test} != *"openmpi-gloo"* ) ]]; then
+    if [[ ! (  ${test} == *"gloo"* && ${test} != *"openmpi-gloo"* ) ]]; then
```
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:309]: same here

**Comment by tgaddair** [comment]
> Do we need to mention Python3 in docs somewhere?

I added a mention in the Install Guide under "requirements".  Might be worth it to put it somewhere on the main README, but couldn't think of a good place.

**Comment by EnricoMi** [comment]
> > Do we need to mention Python3 in docs somewhere?
> 
> I added a mention in the Install Guide under "requirements". Might be worth it to put it somewhere on the main README, but couldn't think of a good place.

I think `install.rst` is good enough, haven't seen that before I send that comment.

**Comment by EnricoMi** [comment]
Are these imports due to Python 2 compatibility?
```
try:
    from shlex import quote
except ImportError:
    from pipes import quote
```

```
try:
    from collections.abc import Iterable
except ImportError:
    from collections import Iterable
```

**Comment by tgaddair** [comment]
> Are these imports due to Python 2 compatibility?
> 
> ```
> try:
>     from shlex import quote
> except ImportError:
>     from pipes import quote
> ```
> 
> ```
> try:
>     from collections.abc import Iterable
> except ImportError:
>     from collections import Iterable
> ```

Good catch, looks like these imports both work in Python 3.6 and above, so I'll update.

**Comment by EnricoMi** [comment]
There is one more Python2 condition in .buildkite/gen-pipeline.sh: https://github.com/horovod/horovod/blob/master/.buildkite/gen-pipeline.sh#L261
